gapis/api/vulkan/looping_vulkan_control_flow_generator.go

```diff
@@ -338,210 +356,258 @@ func newFrameLoop(ctx context.Context, graphicsCapture *capture.GraphicsCapture,
 	}
 }
 
-func (f *frameLoop) Transform(ctx context.Context, cmdId api.CmdID, cmd api.Cmd, out legacyTransformWriter) error {
+func (f *loopingVulkanControlFlowGenerator) TransformAll(ctx context.Context) error { // ALAN: TODO: f -> cf
 
-	// If we're looping only once we can just passthrough commands
-	if f.loopCount == 1 {
-		return out.MutateAndWrite(ctx, cmdId, cmd)
-	}
+	numberOfCmds := f.chain.GetNumOfRemainingCommands()
+	ctx = status.Start(ctx, "Running LoopingVulkanControlFlow <count:%v>", numberOfCmds)
+	defer status.Finish(ctx)
 
-	ctx = log.Enter(ctx, "FrameLoop Transform")
-	log.D(ctx, "FrameLoop: looping from %v to %v. Current CmdID/CmD = %v/%v", f.loopStartIdx, f.loopEndIdx, cmdId, cmd)
-	log.D(ctx, "f.loopTerminated = %v, f.lastObservedCommand = %v", f.loopTerminated, f.lastObservedCommand)
+	var id transform.CommandID
+	defer recoverFromPanic(id)
 
-	// Lets capture and update the last observed frame from f. From this point on use the local lastObservedCommand variable.
-	lastObservedCommand := f.lastObservedCommand
-	f.lastObservedCommand = cmdId
+	subctx := keys.Clone(context.Background(), ctx)
+	for !f.chain.IsEndOfCommands() {
+		id = f.chain.GetCurrentCommandID()
+		if id.GetCommandType() == transform.TransformCommand {
+			cmdID := uint64(id.GetID())
+			if cmdID%100 == 99 {
+				status.UpdateProgress(ctx, cmdID, numberOfCmds)
+			}
+		}
 
-	if lastObservedCommand != api.CmdNoID && lastObservedCommand > api.CmdID.Real(cmdId) {
-		return fmt.Errorf("FrameLoop: expected next observed command ID to be >= last observed command ID")
-	}
+		cmdId := id.GetID()
 
-	// Walk the frame count forwards if we just hit the end of one.
-	if _, ok := cmd.(*VkQueuePresentKHR); ok {
-		f.frameNum++
-	}
+		// Lets capture and update the last observed frame from f. From this point on use the local lastObservedCommand variable.
+		lastObservedCommand := f.lastObservedCommand
+		f.lastObservedCommand = cmdId
+		if cmds, err := f.chain.ProcessNextTransformedCommands(subctx); err == nil {
 
-	// Are we before the loop or just at the start of it?
-	if lastObservedCommand == api.CmdNoID || lastObservedCommand < f.loopStartIdx {
+			log.W(ctx, "FrameLoop: looping from %v to %v. Current CmdID/CmDs = %v/%v", f.loopStartIdx, f.loopEndIdx, id, cmds)
```

Comment:
When I looked into the Log tab, it seems that for a specific CmdId, this log message will only appear once even when looping should be on, I may be missing something here but is it the intended behavior?

Reply:
Yup, it's processing the commands as they appear on the way in to the looping code, not as they are emitted. The actual looping is done by moving the "program counter" in the interpreter anyway so the number of times a command is called is not related to the number of times it is encoded into the bytecode stream.